win32ss/user/ntuser/winpos.c

```diff
@@ -3709,35 +3709,44 @@ NtUserSetWindowPlacement(HWND hWnd,
    TRACE("Enter NtUserSetWindowPlacement\n");
    UserEnterExclusive();
 
-   if (!(Wnd = UserGetWindowObject(hWnd)) ||
-        UserIsDesktopWindow(Wnd) || UserIsMessageWindow(Wnd))
-   {
-      goto Exit; // Return FALSE
-   }
+    Wnd = UserGetWindowObject(hWnd);
+    if (!Wnd)
+        goto Exit; // Return FALSE
+
+    UserRefObjectCo(Wnd, &Ref);
+    if (UserIsDesktopWindow(Wnd) || UserIsMessageWindow(Wnd))
+    {
+        UserDerefObjectCo(Wnd);
+        goto Exit; // Return FALSE
+    }
 
    _SEH2_TRY
    {
       ProbeForRead(lpwndpl, sizeof(WINDOWPLACEMENT), 1);
-      RtlCopyMemory(&Safepl, lpwndpl, sizeof(WINDOWPLACEMENT));
+      Safepl = *lpwndpl;
    }
    _SEH2_EXCEPT(EXCEPTION_EXECUTE_HANDLER)
    {
+      UserDerefObjectCo(Wnd);
```

Comment:
It looks to me that if we were doing the probe and capture of the windowplacement structure first, before checking the window object, this would simplify a bit the code (this would avoid doing that deref call within the exception block).
What do you think? Would this still make the tests succeed?

Reply:
Simplified. Successful.